src/modm/processing/fiber/fiber.hpp

```diff
@@ -16,89 +16,96 @@
+namespace modm
+{
 
 class Fiber;
-// Call to yield execution to another fiber.
-modm_always_inline void yield();
+inline void
+yield();
 
-namespace fiber {
+namespace fiber
+{
 
 class Scheduler;
 
 template<int size>
-class Stack {
-  static_assert(size % 4 == 0, "Stack size must be multiple of 4.");
-  static_assert(size >= 40, "Stack size must at least 40 bytes.");
-  friend class ::modm::Fiber;
-  friend class Scheduler;
+class Stack
+{
+	static_assert(size % 4 == 0, "Stack size must be multiple of 4.");
+	static_assert(size >= 40, "Stack size must at least 40 bytes.");
+	friend class ::modm::Fiber;
+	friend class Scheduler;
+
 public:
-  Stack() = default;
-  Stack(const Stack&) = delete;
+	Stack() = default;
+	Stack(const Stack&) = delete;
 
 private:
-public:
-  modm_stack_t memory_[size / sizeof(modm_stack_t)];
+	modm_aligned(8)
+	modm_stack_t memory_[size / sizeof(modm_stack_t)];
 };
 
 } // namespace fiber
 
-class Fiber {
-  friend class fiber::Scheduler;
-  friend void yield();
+class Fiber
+{
+	friend class fiber::Scheduler;
+	friend void yield();
  public:
-  template<int size>
-  Fiber(fiber::Stack<size>& stack, void(*f)());
-  void dumpStack();
+	template<int size>
+	Fiber(fiber::Stack<size>& stack, void(*f)());
+
+	inline void
+	dumpStack();
 
 protected:
-  modm_always_inline
-  void jump(Fiber& other) {
-    modm_jumpcontext(&ctx_, other.ctx_);
-  }
+	inline void
+	jump(Fiber& other)
+	{ modm_jumpcontext(&ctx_, other.ctx_); }
+
+private:
+	Fiber() = default;
+	Fiber(const Fiber&) = delete;
+
+	// Removes the current fiber from the scheduler and yields execution. This is called on fiber
+	// return.
+	static inline void
+	done();
 
 private:
-  Fiber() = default;
-  Fiber(const Fiber&) = delete;
-  modm_context ctx_;
-  modm_stack_t stack_;
-  Fiber* next_;
-
-  // Removes the current fiber from the scheduler and yields execution. This is called on fiber
-  // return.
-  modm_always_inline static void done();
+	modm_context ctx_;
+	modm_stack_t stack_;
+	Fiber* next_;
 };
 
-namespace fiber {
+namespace fiber
+{
 
-class Scheduler {
-  friend class modm::Fiber;
-  friend void modm::yield();
+class Scheduler
+{
+	friend class modm::Fiber;
+	friend void modm::yield();
 public:
-  constexpr Scheduler() = default;
-  /* Should be called by the main() function. */
-  modm_always_inline
-  void start();
+	Scheduler() = default;
+
+	inline void
+	start();
+
 protected:
-  modm_always_inline
-  void registerFiber(Fiber*);
-  modm_always_inline
-  Fiber* currentFiber() { return current_fiber_; }
+	inline void
+	registerFiber(Fiber*);
+
+	inline Fiber*
+	currentFiber() { return current_fiber_; }
+
  private:
-  Scheduler(const Scheduler&) = delete;
-  Fiber* current_fiber_ = nullptr;
+	Scheduler(const Scheduler&) = delete;
+	Fiber* current_fiber_ = nullptr;
 };
 
-static Scheduler scheduler_instance_;
-
-modm_always_inline
-static Scheduler& scheduler() {
-  return scheduler_instance_;
-}
+static inline Scheduler scheduler;
```

Comment:
Not a big deal, but I believe this is wrong because initialization order of static variables across compilation units is undefined.

That said, we are probably safe because Scheduler can be compile time instantiated.

Reply:
I actually wanted to remove the Scheduler completely and only have function calls, but lost motivation in the middle ;-P

Reply:
Oh, I see, you wanted static function initialization, but then you need to move the `static Scheduler scheduler_instance_` inside the function I think? I added the guards for that in #346.

Reply:
Yeah you are right, delayed initialization was my intention. We still need somewhere to store the ready list and the current running fiber, given that this is not FreeRTOS I decided to use C++ object ;)

Btw I was considering permanently storing the scheduler address in R9, if I understand the ABI correctly then that register is intended for RTOS context state. It might give us a few tens of nS lower context switch latency

Reply:
Oh interesting. I was aware of R9's role in PIC and TLS, but didn't know you could use it for anything you want.

> The role of register r9 is platform specific. A virtual platform may assign any role to this register and must document this usage. For example, it may designate it as the static base (SB) in a position-independent data model, or it may designate it as the thread register (TR) in an environment with thread-local storage. The usage of this register may require that the value held is persistent across all calls. A virtual platform that has no need for such a special register may designate r9 as an additional callee-saved variable register, v6.

But it's not clear to me how to enable this function in the toolchain? I only found these two options in the [ARM Options](https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html).

> -msingle-pic-base
> Treat the register used for PIC addressing as read-only, rather than loading it in the prologue for each function. The runtime system is responsible for initializing this register with an appropriate value before execution begins.
>
> -mpic-register=reg
> Specify the register to be used for PIC addressing. For standard PIC base case, the default is any suitable register determined by compiler. For single PIC base case, the default is ‘R9’ if target is EABI based or stack-checking is enabled, otherwise the default is ‘R10’.

It may be useful to use R11 for this purpose, since that makes the push/pop code smaller.